content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+
+Format of a refactoring:
+- They have a name
+  - Is important to building a vocabulary of refactoring.
+- They have a motivation
+  - Describes why the refactoring should be done and describes circumstances in which it shouldn’t be done.
+- They have a mechanic
+  - Step-by-step description of how to carry out the refactoring.
+
+Mechanics of a refactoring:
+- Each step in the refactoring is as small as possible.
+- Test after every step. 
+- In practice, if you want to take larger steps, you can, but if the tests fail, then you should go backwards and take smaller steps.
+
+In the book they are presented within 6 different categories which may not be worth it to explain them in this article. Here you can find a [catalog of refactoring](https://refactoring.com/catalog/).
+
+**Conclusion: There is no conclusion here.**
+
+## 6) SOME EXAMPLES
+If you read up to here, first thank you, and second, now you have to finish it.
+
+I will present just 2 simple refactoring as examples. I don’t think the purpose of this article should be for you to learn many of them or learning some complex ones. I believe that is something you can learn by yourself in somewhere else, but you will do it only if you feel enough motivation to do it.
+
+**First Example - Decompose Conditional Expression:**
+You have a complicated conditional (if-then-else) statement.
+→ Extract methods from the condition, “then” part, and “else” part.
+
+![Decompose Conditional Expression Example](images/decompose_conditional_expression.png)
+
+- Motivation: Conditional logic adds complexity to the code. You can make your intention clearer by decomposing it and replacing chunks of code with a method call named after the intention of that block of code.
+- Mechanic:
+  1.  Extract the conditions into their own methods.
+  2.  Test.
+  3.  Extract the actions inside the conditional code into their own methods.
+  4.  Test.
+
+**Second Example - Extract Class:**
+You have one class doing work that should be done by two.
+→ Create a new class and move the relevant fields and methods from the old class into the new class.
+
+![Exctact Class Example](images/extract_class_1.png)
+
+![Exctact Class Example](images/extract_class_2.png)
+
+![Exctact Class Example](images/extract_class_3.png)
+
+- Motivation: In practice, classes grow over time. One class with lots of methods and data is too big to understand easily, so it becomes hard to maintain.
+- Mechanic:
+  1. Decide how to split the responsibilities of the class.
+  2. Create a new class to express the split-off responsibilities. If the responsibilities of the old class no longer match its name, rename the old class.
+  3. Make a link from the old to the new class.
+  4. Use Move Field on each field you want to move.
+  5. Test after each move.
+  6. Use Move Method to move methods over from the old to the new. Start with lower-level methods and build to the higher level.
```

Comment:
```suggestion
  6. Use Move Method to move methods over from the old to the new. Start with lower-level methods and build to a higher level.
```

Reply:
Not sure on this one. I will ask an english speaker. Thanks.